src/plugins/discover/public/application/main/components/top_nav/discover_topnav.tsx

```diff
@@ -186,10 +186,13 @@ export const DiscoverTopNav = ({
   );
 
   const onEditDataView = async (editedDataView: DataView) => {
-    if (!editedDataView.isPersisted()) {
-      await updateAdHocDataViewId(editedDataView);
+    if (editedDataView.isPersisted()) {
+      // Clear the current data view from the cache and create a new instance
+      // of it, ensuring we have a new object reference to trigger a re-render
+      dataViews.clearInstanceCache(editedDataView.id);
+      stateContainer.actions.setDataView(await dataViews.create(editedDataView.toSpec(), true));
```

Comment:
Did you try using `dataViews.get` instead of `dataViews.create`? if so, what problem did you run into?

Reply:
I didn't and I can't imagine there'd be an issue using it. I just figured we'd prefer to avoid the additional request if it isn't necessary (especially with the `dataViews.refreshFields` call that `data_view_flyout_content_container.tsx` already makes after a save now), but I can definitely try with `dataViews.get` instead if preferred.

Reply:
I have two thoughts looking at this code -

a) Is the data views api failing the api consumer? 

b) Is the need to get a new object causing us to do funny things in our code?

In this case I think the answer is `b` although I'm not sure what to do about it at this very moment. It seems like a flaw in either our state modeling or tooling. 

Looking at the data views code, I'm not sure that create and get are functionally different for this use case.

Reply:
> In this case I think the answer is `b` although I'm not sure what to do about it at this very moment. It seems like a flaw in either our state modeling or tooling.

it's more a reacty thing, as mentioned in the comment it needs a new object reference to trigger re-rendering in components that use the data view internal state. Now that we allow data view edits, we might think how we could improve the handling of this (but that's a refactoring better to handle in a follow up )

Some possible ways to handle this differently:
* Handle the internal state of the data view differently, e.g. like this`{version: 1, dataView: DataView}`, so on edit the version is incremented and a new object is set in internal state
* Provide a `clone` method to a data view, returning the same data view but with a different object ref
* Generally, when a data view is updated, the data views service could return a new data view object
* Some other way?


Generally in React, not mutating objects saves a lot of headache, in my experience

Reply:
I updated the call from `.create` to `.get`. In testing I found this does add additional requests for `/api/saved_objects/_bulk_resolve` and `/api/index_patterns/_fields_for_wildcard` compared to the `.create` approach which needs no additional fetching, but as long as CI passes, I'm fine with merging as is if it's considered the more correct way to use the API.

I should point out though that we now call `/api/index_patterns/_fields_for_wildcard` twice after saving a data view due to this change + the changes introduced in #149861 (which probably aren't technically necessary for the orignal issue anymore since we're now reloading the whole data view, but it does still solve the general use case for the data view flyout). This might not be important in practice given that data views are likely edited infrequently, but I figured I should point it out in light of the recent discussions around the performance of data view field list loading.

Reply:
@davismcphee Apologies for that, I must have misread the data views service code. I'm ambivalent about how we should leave this code, you can decide. In the near future I hope we can make `dataViews.get` more efficient in which case we'd want to revisit this.

Reply:
No worries, I decided to revert the commit to save a couple hops to the server for now, but I agree that a better long term solution would be preferable.